fsi/mapping.go

```diff
@@ -1,436 +1,122 @@
-		case "json":
-			bodyFilename = "body.json"
-		default:
-			return fmt.Errorf(`unknown body format: "%s"`, bodyFormat)
-		}
-		err = ioutil.WriteFile(filepath.Join(dirPath, bodyFilename), data, os.ModePerm)
-		if err != nil {
-			return err
-		}
+// WriteComponent writes the component with the given name to the directory
+func WriteComponent(comp component.Component, name string, dirPath string) error {
+	aComp := comp.Base().GetSubcomponent(name)
+	if aComp == nil {
+		return nil
 	}
+	return aComp.WriteTo(dirPath)
+}
 
-	// Dataset (everything else).
-	ds.DropDerivedValues()
-	ds.DropTransientValues()
-	// TODO(dlong): Should more of these move to DropDerivedValues?
-	ds.Qri = ""
-	ds.Peername = ""
-	ds.PreviousPath = ""
-	if !ds.IsEmpty() {
-		data, err := json.MarshalIndent(ds, "", " ")
-		if err != nil {
-			return err
-		}
-		err = ioutil.WriteFile(filepath.Join(dirPath, "dataset.json"), data, os.ModePerm)
-		if err != nil {
-			return err
-		}
+// DeleteComponent deletes the component with the given name from the directory
+func DeleteComponent(comp component.Component, name string, dirPath string) error {
+	aComp := comp.Base().GetSubcomponent(name)
+	if aComp == nil {
+		return nil
 	}
-
-	return nil
+	return aComp.RemoveFrom(dirPath)
 }
 
 // DeleteDatasetFiles removes mapped files from a directory. if the result of
 // deleting all files leaves the directory empty, it will remove the directory
 // as well
-func DeleteDatasetFiles(dirPath string) (removed map[string]FileStat, err error) {
-	_, mapping, problems, err := ReadDir(dirPath)
+func DeleteDatasetFiles(dirPath string) error {
+	components, err := component.ListDirectoryComponents(dirPath)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	removed = map[string]FileStat{}
-	for component, stat := range mapping {
-		// ignore not found errors. multiple components can be specified in the
-		// same dataset file, creating multiple remove attempts to the same path
-		if err := os.Remove(stat.Path); err != nil && !os.IsNotExist(err) {
-			return nil, err
-		}
-		removed[component] = stat
-	}
+	allComponents := append(component.AllSubcomponentNames(), "dataset")
 
-	// delete files even if they have problems parsing
-	for component, stat := range problems {
-		// TODO (b5): mapping returns absolute paths in FileStat, problems returns
-		// relative paths. We should pick one & go with it. I vote absolute
-		path := filepath.Join(dirPath, stat.Path)
+	for _, compName := range allComponents {
+		subc := components.Base().GetSubcomponent(compName)
+		if subc == nil {
+			continue
+		}
 		// ignore not found errors. multiple components can be specified in the
 		// same dataset file, creating multiple remove attempts to the same path
-		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
-			return nil, err
-		}
-		removed[component] = FileStat{
-			Path:  path,
-			Mtime: stat.Mtime,
+		if err := os.Remove(subc.Base().SourceFile); err != nil && !os.IsNotExist(err) {
```

Comment:
it should be a long term goal of this package to _only_ interact with a qfs.Filesystem, pushing all `os` package interaction into qfs land. We're definitely not there today.